billing-report-agent/src/main/java/com/epam/pipeline/billingreportagent/service/impl/converter/RunToBillingRequestConverter.java

```diff
@@ -0,0 +1,172 @@
+
+        final Map<LocalDate, PipelineRunBillingInfo> reports = new HashMap<>();
+        activePeriods.forEach(period -> {
+            for (int i = 0; i < period.size() - 1; i++) {
+                final LocalDateTime start = period.get(i).getTimestamp();
+                final LocalDateTime end = period.get(i + 1).getTimestamp();
+                final Long dailyCost = calculateCostsForPeriod(start, end, pricePerHour);
+                addDailyCostReport(reports, start.toLocalDate(), dailyCost, ChronoUnit.MINUTES.between(start, end), run);
+            }
+        });
+        return new ArrayList<>(reports.values());
+    }
+
+    private void adjustStatuses(final List<RunStatus> statuses, final LocalDateTime syncStart, final Long runId) {
+        statuses.sort(Comparator.comparing(RunStatus::getTimestamp));
+        final RunStatus lastStatus = statuses.get(statuses.size() - 1);
+        if (lastStatus.getStatus() == TaskStatus.RUNNING) {
+            statuses.add(new RunStatus(null, null, syncStart));
+        } else if (statuses.size() > 1) {
+            final RunStatus status = statuses.get(statuses.size() - 2);
+            if (status.getStatus() != TaskStatus.RUNNING) {
+                statuses.remove(statuses.size() - 1);
+            }
+        }
+        if (statuses.size() % 2 != 0) { // no RUNNING states for given run
+            log.warn("Can't adjusting run {} statuses, it has no RUNNING states in history!", runId);
+            statuses.clear();
+        }
+    }
+
+    private List<List<RunStatus>> calculateActivityPeriods(final List<RunStatus> statuses) {
+        final List<List<RunStatus>> activePeriods = new ArrayList<>();
+        for (int i = 0; i < statuses.size(); i += 2) {
+            final RunStatus currentStatus = statuses.get(i);
+            final RunStatus nextStatus = statuses.get(i + 1);
+            final List<RunStatus> activePeriod = new ArrayList<>();
+            activePeriod.add(currentStatus);
+            LocalDateTime nextTimePoint = currentStatus.getTimestamp()
+                .toLocalDate()
+                .plusDays(1)
+                .atTime(LocalTime.MIDNIGHT);
+            while (nextTimePoint.isBefore(nextStatus.getTimestamp())) {
+                activePeriod.add(new RunStatus(null, null, nextTimePoint));
+                nextTimePoint = nextTimePoint.plusDays(1);
+            }
+            activePeriod.add(nextStatus);
+            activePeriods.add(activePeriod);
+        }
+        return activePeriods;
+    }
+
+    private void addDailyCostReport(final Map<LocalDate, PipelineRunBillingInfo> reports,
+                                    final LocalDate date,
+                                    final Long cost,
+                                    final Long duration,
+                                    final EntityContainer<PipelineRun> run) {
+        final PipelineRunBillingInfo billing = reports.get(date);
+        if (billing != null) {
```

Comment:
Can we replace with `Optional`?

Reply:
Removed due to changing approach of active period evaluating